From 1e71c0d1d30e28e10d01ec6ec6f5ca92c14691e1 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Sat, 8 Feb 2025 16:17:36 +0800
Subject: [PATCH 101/101] desktop-shell: Support setting default window
 position

Tested on RK3588 EVB with:
WESTON_DEFAULT_POSITION=100,200 weston
WESTON_DEFAULT_POSITION=center weston
WESTON_DEFAULT_POSITION=left-top weston
WESTON_DEFAULT_POSITION=left-bottom weston
WESTON_DEFAULT_POSITION=right-top weston
WESTON_DEFAULT_POSITION=right-bottom weston

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 desktop-shell/shell.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index a34ac37db..0a7e47b58 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -4171,6 +4171,7 @@ weston_view_set_initial_position(struct weston_view *view,
 	struct weston_seat *seat;
 	pixman_rectangle32_t area;
 	struct weston_coord_global pos;
+	const char *buf;
 
 	if (view->has_position) {
 		pos.c = weston_coord(view->geometry.pos_offset.x,
@@ -4232,6 +4233,30 @@ weston_view_set_initial_position(struct weston_view *view,
 	if (range_y > 0)
 		y += random() % range_y;
 
+	buf = getenv("WESTON_DEFAULT_POSITION");
+	if (buf) {
+		int tmp_x, tmp_y;
+		if (sscanf(buf, "%d,%d", &tmp_x, &tmp_y) == 2) {
+			x = tmp_x;
+			y = tmp_y;
+		} else if (!strcmp(buf, "center")) {
+			x = area.x + range_x / 2;
+			y = area.y + range_y / 2;
+		} else if (!strcmp(buf, "left-top")) {
+			x = area.x;
+			y = area.y;
+		} else if (!strcmp(buf, "left-bottom")) {
+			x = area.x;
+			y = area.y + range_y;
+		} else if (!strcmp(buf, "right-top")) {
+			x = area.x + range_x;
+			y = area.y;
+		} else if (!strcmp(buf, "right-bottom")) {
+			x = area.x + range_x;
+			y = area.y + range_y;
+		}
+	}
+
 	pos.c = weston_coord(x, y);
 	weston_view_set_position(view, pos);
 }
-- 
2.20.1

